### PR TITLE
Implemented passing unmanaged views

### PR DESCRIPTION
**Pull request by helloworld922**
For the output reverse transpose this will try to reshape a given unmanaged buffer if it fits, otherwise throws an exception.

### CODE REVIEW

**Review by yasahi-hpc — CHANGES_REQUESTED** [review]

File: common/src/KokkosFFT_transpose.hpp [hunk 1]
```diff
@@ -86,6 +86,31 @@ auto get_map_axes(const ViewType& view, int axis) {
   return get_map_axes(view, axis_type<1>({axis}));
 }
 
+template <class InViewType, class OutViewType, std::size_t DIMS>
+void _prep_transpose_view(InViewType& in, OutViewType& out,
+                          axis_type<DIMS> _map) {
```
Comment (yasahi-hpc) [inline common/src/KokkosFFT_transpose.hpp:91]: Could you please add an unit-test for this function, for example in common/unit_test/Test_Transpose.cpp. An exception test for insufficient memory case would be quite helpful.

File: examples/03_NDFFT/03_NDFFT.cpp
[the remark this answers is not in the document]
Comment (helloworld922) [reply examples/03_NDFFT/03_NDFFT.cpp:1]: added as 07_unmanaged_views where a shared allocation is used for the fft/ifft input and output and a separate managed view for the intermediate FFT.

File: common/src/KokkosFFT_transpose.hpp (hunk 1 above)
Comment (helloworld922) [reply common/src/KokkosFFT_transpose.hpp:91]: Added in Test_prep_transpose_view.cpp

This tests allocating/no allocating cases for managed views, and no reshape/reshape success/reshape failure for unmanaged views.

File: examples/03_NDFFT/03_NDFFT.cpp
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply examples/03_NDFFT/03_NDFFT.cpp:1]: Thanks. Looks good

**Review by yasahi-hpc — APPROVED** [review]

File: common/unit_test/Test_prep_transpose_view.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+      map[i]              = i;
+    }
+
+    LayoutType layout_orig;
+    layout_orig.dimension[0] = 1;
+    for (int i = 0; i < DIMS; ++i) {
+      layout_orig.dimension[0] *= 5;
+    }
+    InManagedViewType in("in", layout);
+    OutManagedViewType out("out", layout_orig);
```
Comment (yasahi-hpc) [inline common/unit_test/Test_prep_transpose_view.cpp:121]: Seems like we need a rank-1 version of `OutManagedViewType` for this operation.

File: examples/07_unmanaged_views/07_unmanaged_views.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    shape_type<3> shape;
+    shape[0] = n0;
+    shape[1] = n1;
+    shape[2] = n2;
+
+    // any combination of inputs/outputs can be managed or unmanaged views
+    // should work on all functions in any number of dimensions
+    // 3D C2C FFT (Forward and Backward)
+    
+    // combined storage buffer for xc2c and xc2c_inv
```
Comment (yasahi-hpc) [inline examples/07_unmanaged_views/07_unmanaged_views.cpp:41]:
```suggestion
    // any combination of inputs/outputs can be managed or unmanaged views
    // should work on all functions in any number of dimensions
    // 3D C2C FFT (Forward and Backward)

    // combined storage buffer for xc2c and xc2c_inv
```

File: common/unit_test/Test_prep_transpose_view.cpp (hunk 2 above)
Comment (helloworld922) [reply common/unit_test/Test_prep_transpose_view.cpp:127]: Added, as well as properly specify all extents for higher dimensions to 1.

Comment (yasahi-hpc) [reply common/unit_test/Test_prep_transpose_view.cpp:127]: Sure

File: common/unit_test/Test_prep_transpose_view.cpp [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  {
+    LayoutType layout;
+    KokkosFFT::axis_type<DIMS> map;
+    for (int i = 0; i < DIMS; ++i) {
+      layout.dimension[i] = 5;
+      map[i]              = i;
+    }
+
+    LayoutType layout_orig;
+    layout_orig.dimension[0] = 1;
```
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply common/unit_test/Test_prep_transpose_view.cpp:145]: Need to fix this also

**Review by yasahi-hpc — APPROVED** [review]

**Pull request merged**